src/viadot/sources/sftp.py

```diff
@@ -0,0 +1,296 @@
+
+import itertools
+import re
+import time
+from collections import defaultdict
+from io import BytesIO, StringIO
+from pathlib import Path
+from stat import S_ISDIR
+
+import pandas as pd
+import paramiko
+from paramiko.sftp import SFTPError
+from pydantic import BaseModel
+
+from viadot.config import get_source_credentials
+from viadot.exceptions import CredentialError
+from viadot.sources.base import Source
+from viadot.utils import add_viadot_metadata_columns
+
+
+class SftpCredentials(BaseModel):
+    """Checking for values in SFTP credentials dictionary.
+
+    Two key values are held in the Salesforce connector:
+        - hostname: The direction for the SFTP.
+        - username: The unique name for SFTP connection.
+        - password: The unique passwrod for SFTP connection.
+        - port: Number from which the connection will be done.
+        - rsa_key: The Company RSA Key.
+
+    Args:
+        BaseModel (pydantic.main.ModelMetaclass): A base class for creating
+            Pydantic models.
+    """
+
+    hostname: str
+    username: str
+    password: str
+    port: int
+    rsa_key: str
+
+
+class SftpConnector(Source):
+    """Class implementing a SFTP server connection."""
+
+    def __init__(
+        self,
+        *args,
+        credentials: SftpCredentials | None = None,
+        config_key: str = "sftp",
+        **kwargs,
+    ):
+        """Create an instance of SFTP.
+
+        Args:
+            credentials (SftpCredentials, optional): SFTP credentials. Defaults to None.
+            config_key (str, optional): The key in the viadot config holding relevant
+                credentials. Defaults to "sftp".
+
+        Notes:
+            self.conn is paramiko.SFTPClient.from_transport method that contains
+            additional methods like get, put, open etc. Some of them were not
+            implemented in that class. For more check documentation
+            (https://docs.paramiko.org/en/stable/api/sftp.html).
+
+            sftp = SftpConnector()
+            sftp.conn.open(filename='folder_a/my_file.zip', mode='r')
+
+        Raises:
+            CredentialError: If credentials are not provided in local_config or
+                directly as a parameter.
+        """
+        credentials = credentials or get_source_credentials(config_key) or None
+
+        if credentials is None:
+            message = "Missing credentials."
+            raise CredentialError(message)
+        self.credentials = credentials
+
+        validated_creds = dict(SftpCredentials(**credentials))
+        super().__init__(*args, credentials=validated_creds, **kwargs)
+
+        self.conn = None
+        self.hostname = validated_creds.get("hostname")
+        self.username = validated_creds.get("username")
+        self.password = validated_creds.get("password")
+        self.port = validated_creds.get("port")
+        self.rsa_key = validated_creds.get("rsa_key")
+
+    def _getfo_file(self, file_name: str) -> BytesIO:
```

Comment:
Pls use full names. I suppose this is `_get_file_object()`?

Reply:
changed.